doc/developer/design/20230728_left_join_stack_consolidation.md

```diff
@@ -0,0 +1,160 @@
+         Project (#0, #1, #0, #2)            +
+           Get l0                            + // 1, 1
+     cte l0 =                                +
+       Project (#0, #1, #3)                  +
+         Join on=(#0 = #2) type=differential + // 1, 1
+           ArrangeBy keys=[[#0]]             +
+             Filter (#0) IS NOT NULL         +
+               Get materialize.public.foo    +
+           ArrangeBy keys=[[#0]]             +
+             TopK group_by=[#0] limit=1      +
+               Filter (#0) IS NOT NULL       +
+                 Get materialize.public.bar  + // 1, 1
+                                             +
+ Source materialize.public.bar               +
+   filter=((#0) IS NOT NULL)                 +
+ Source materialize.public.more1             +
+   filter=((#0) IS NOT NULL)                 +
+ Source materialize.public.more2             +
+   filter=((#0) IS NOT NULL)                 +
+```
+For each of the 3 LEFT JOINs, the plan has one `Union-Negate`: In `l1`, `l3`, and the `Let` body. The problem is that there is a dataflow path that has no consolidation between these Union-Negates: For example, the problematic path between the first two Union-Negates is as follows: We start from the Union-Negate in `l1`, then we have a Map, another Union (with no Negate), then we go to the first usage of `l1` (the second usage is fine, since it goes into a Join), which goes into the Union-Negate in `l3`.
+
+The Union-Negate pattern computes the outer part of a LEFT JOIN: the negated input to these unions is the collection of matched records. So, we have the most waste when all records have a match, in which case a consolidation would eliminate all records coming out from the Union-Negate. We can see that if we don't consolidate, then the record count increases by `2N`, where `N` is the size of one input, whereas if we consolidate, then record counts stay at `N`. Note that the increase is _not_ exponential (we have N, 3N, 5N, 7N, ...), but it will still get quite bad for big stacks. ([We've seen several big stacks](https://www.notion.so/materialize/Left-Join-Feedback-1d921cbe0296431b96b97b2480808cf9?d=e38c0c85241c48a98e8414f75e2c37e0). One customer [has a stack of 28 LEFT JOINs in one view, and 12 and 13 in two other views](https://materializeinc.slack.com/archives/C05GRCC4K1C/p1690293441722169).)
+
+Note that the problem occurs regardless of whether SemijoinIdempotence kicks in.
+
+# Explanation
+[explanation]: #explanation
+
+As mentioned above, our proposed solution is to simply perform a consolidation immediately after each of the Union-Negates. More specifically, during the MIR-to-LIR lowering of a Union, we'll check if the Union has an input that has a Negate at the top, and if yes, then we set a `consolidate` flag on the Union, which will make the rendering insert a `consolidate` call on the Union's result. This will immediately make the negated records cancel out.
```

Comment:
Doing this in lowering sounds pretty manageable, but I just wanted to call out another option: We can have a refinement written in iterative style, similarly to what is done [here](https://github.com/MaterializeInc/materialize/blob/1b47c5f0cb8b021a253f1bf54eeeab500286abf6/src/compute-client/src/plan/mod.rs#L1837-L1887), that matches `Plan::Union`s that have a `Plan::Negate` child and flip the flag. This can be helpful to avoid cluttering lowering and allowing you to trace this setting too, but it requires another pass.

Reply:
Hmm, right, it seems better to do it as a refinement. I've changed the text.